cleanlab_studio/utils/data_enrichment/enrichment_utils.py

```diff
@@ -0,0 +1,121 @@
+
+
+def get_prompt_outputs(
+    studio: Studio, prompt: str, data: pd.DataFrame, **kwargs: Any
+) -> List[TLMResponse | None]:
+    """Returns the outputs of the prompt for each row in the dataframe."""
+    tlm = studio.TLM(**kwargs)
+    formatted_prompts = data.apply(lambda x: prompt.format(**x), axis=1).to_list()
+    outputs = tlm.try_prompt(formatted_prompts)
+    return outputs
+
+
+def extract_df_subset(
+    df: pd.DataFrame, subset_indices: Union[Tuple[int, int], List[int], None, range]
+) -> pd.DataFrame:
+    """Extract a subset of the dataframe based on the provided indices. If no indices are provided, the entire dataframe is returned. Indices can be range or specific row indices."""
+    if subset_indices is None:
+        print(
+            "Processing your full dataset since `subset_indices` is None. This may take a while. Specify this argument to get faster results for a subset of your dataset."
+        )
+        return df
+    if isinstance(subset_indices, range):
+        subset_indices = subset_indices
+    if isinstance(subset_indices, tuple):
+        subset_indices = range(*subset_indices)
+    subset_df = df.iloc[subset_indices].copy()
+    return subset_df
+
+
+def get_compiled_regex_list(
+    regex: Union[str, re.Pattern[str], List[re.Pattern[str]]]
+) -> List[re.Pattern[str]]:
+    """Compile the regex pattern(s) provided and return a list of compiled regex patterns."""
+    if isinstance(regex, str):
+        return [re.compile(rf"{regex}")]
+    elif isinstance(regex, re.Pattern):
+        return [regex]
+    elif isinstance(regex, list):
+        return regex
+    else:
+        raise ValidationError(
+            "Passed in regex can only be type one of: str, re.Pattern, or list of re.Pattern."
+        )
+
+
+def get_regex_match(
+    response: str, regex_list: List[re.Pattern[str]], disable_warnings: bool = False
+) -> Union[str, None]:
+    """Extract the first match from the response using the provided regex patterns. Return first match if multiple exist.
+    Note: This function assumes the regex patterns each specify exactly 1 group that is the match group using '(<group>)'."""
+    for regex_pattern in regex_list:
+        pattern_match = regex_pattern.match(response)
+        if pattern_match:
+            return pattern_match.group(1)
+    if not disable_warnings:
+        warnings.warn(
+            f"Your provided regex: {str(regex_list)} did not match a single LLM output across the dataset. This message is simply to inform you that the regex is not having any effect."
+        )
+    return None
+
+
+def get_optimized_prompt(prompt: str, return_values: Optional[List[str]] = None) -> str:
+    """Optimize the prompt by ammending original.
+    Adds a pre-prompt message if return_values are provided. This will help the LLM understand it's response must exactly match one of the return values."""
+
+    if return_values is not None:
+        string_return_values = str(return_values).replace("'", "")
```

Comment:
I believe this should look like this?

[a, b, c]

and not like this:

a, b, c

so that there is a clear delimeter around this list?